packages/paima-sdk/paima-utils/src/config/loading.ts

```diff
@@ -0,0 +1,154 @@
+import * as fs from 'fs/promises';
+import YAML from 'yaml';
+import type { Static, TSchema } from '@sinclair/typebox';
+import type { ValueErrorType } from '@sinclair/typebox/value';
+import { Value } from '@sinclair/typebox/value';
+import { Type } from '@sinclair/typebox';
+import { ENV, doLog } from '../index.js';
+
+export enum ConfigNetworkType {
+  EVM = 'evm-main',
+  EVM_OTHER = 'evm-other',
+  CARDANO = 'cardano',
+}
+
+export type EvmConfig = Static<typeof EvmConfigSchema>;
+
+export const EvmConfigSchema = Type.Object({
+  chainUri: Type.String({ default: '' }),
+  chainId: Type.Number({ default: 0 }),
+  chainExplorerUri: Type.String({ default: '' }),
+  chainCurrencyName: Type.String({ default: 'UNKNOWN_CURRENCY_NAME' }),
+  chainCurrencySymbol: Type.String({ default: 'NONAME' }),
+  chainCurrencyDecimals: Type.Number({ default: 0 }),
+  // TODO: this depends on the settings actually, but hard to express that
+  //ENV.BLOCK_TIME - 0.1
+  blockTime: Type.Number({ default: 4 }),
+  pollingRate: Type.Number({ default: 4 - 0.1 }),
+  deployment: Type.String({ default: '' }),
+  paimaL2ContractAddress: Type.RegExp(/^0x[0-9a-fA-F]{40}$/i),
+  funnelBlockGroupSize: Type.Number({ default: 100 }),
+});
+
+export const CardanoConfigSchema = Type.Object({
+  carpUrl: Type.String(),
+  network: Type.String(),
+  confirmationDepth: Type.Number(),
+});
+
+export type CardanoConfig = Static<typeof CardanoConfigSchema>;
+
+export const TaggedConfig = <T extends boolean>(T: T) =>
+  Type.Union([
+    Type.Intersect([
+      T ? EvmConfigSchema : Type.Partial(EvmConfigSchema),
+      Type.Required(
+        Type.Object({
+          type: Type.Union([
+            Type.Literal(ConfigNetworkType.EVM),
+            Type.Literal(ConfigNetworkType.EVM_OTHER),
+          ]),
+        })
+      ),
+    ]),
+    Type.Intersect([
+      CardanoConfigSchema,
+      Type.Object({ type: Type.Literal(ConfigNetworkType.CARDANO) }),
+    ]),
+  ]);
+
+export const BaseConfig = <T extends boolean>(T: T) => Type.Record(Type.String(), TaggedConfig(T));
+
+export const BaseConfigWithoutDefaults = BaseConfig(false);
+export const BaseConfigWithDefaults = BaseConfig(true);
+export const FullEvmConfig = EvmConfigSchema;
+
+const evmConfigDefaults = (blockTime: number | undefined) => ({
+  chainUri: '',
+  chainId: 0,
+  chainExplorerUri: '',
+  chainCurrencyName: 'UNKNOWN_CURRENCY_NAME',
+  chainCurrencySymbol: 'NONAME',
+  chainCurrencyDecimals: 0,
+  blockTime: 4,
```

Comment:
strange how block time is a parameter, yet here we just set it to 4. I'm not sure why we would even want these defaults instead of just erroring though since other than `chainExplorerUri` and `pollingRate` these are all required (also fyi `deployment` is deprecated)

Reply:
Well, but all of these things currently have defaults... So I just kept that behavior. I took the defaults from the `ENV` class.